### DISCUSSION TITLE
There doesn't seem to be a network

### DISCUSSION THREAD

**Discussion by kkrankall** [CharlesGodwin/pymagnum#37] [Q&A] [answered]
Charles, First off thank you for writing this. I'm trying to get mine to work.

I'm using a Pi 4 Model B with a Waveshare RS485 CAN HAT for Raspberry Pi and din rail unit using the terminals marked 2 and 5. Per your instructions.

I have disabled BT on the config.txt

```
enable_uart=1

# Disable Bluetooth
dtoverlay=disable-bt
```

When I run 
`pi@magnumpi:~ $ python -m serial.tools.list_ports`
I get this back.
```
/dev/ttyAMA0        
1 ports found
```

When I run
`magtest --device /dev/ttyAMA0`
I get this back
```
Magnum Test Version:2.0
Options:device='/dev/ttyAMA0', packets=50, timeout=0.005, trace=False, cleanpackets=True, log=False
/dev/ttyAMA0 There doesn't seem to be a network
Error detected attempting to read network data - test failed.
```

I am using port P5 on my ME ARTR with the phone splitter shared with my BMK. 

I would love to get this up and running and get data into my HASS

**Comment by CharlesGodwin** [comment]
First, try using /dev/ttyS0 (some of the waveshare HAT do this)

If that doesn't do it

Try this,

- sudo raspi-config
- select option 3 - Interface Options
- select option I6 - Serial Port
- select No to Would you like a login shell to be accessible over serial?
- select yes to Would you like the serial port hardware to be enabled?

This will require a reboot

Then repeat your tests. on /dev/ttyS0 and /dev/ttyAMA0

**Reply by kkrankall** [reply]
Wow thanks for the fast response.

```
pi@magnumpi:~ $ magtest --device /dev/ttyS0
Magnum Test Version:2.0
Options:device='/dev/ttyS0', packets=50, timeout=0.005, trace=False, cleanpackets=True, log=False
/dev/ttyS0 Could not configure port: (5, 'Input/output error')
Error detected attempting to read network data - test failed.
```

I already configured the interface options/ serial port earlier when troubleshooting. Didn't change anything.

**Comment by CharlesGodwin** [comment]
I'm at the end of my day.

All I can suggest before morning is to fall back to my older release 1.1.3

Uninstall pymagnum

sudo pip uninstall pymagnum

Then install the older version

sudo pip install pymagnum==1.1.3

I'll try to test with a HAT in the morning

Charles
________________________________

**Reply by kkrankall** [reply]
Thank you. Will give it a try

**Comment by CharlesGodwin** [comment]
I re-installed my WaveShare RS485 HAT this morning. Once I got the wire polarity right. I got results. see attached
[magtest_2022-06-13_093925.txt](https://github.com/CharlesGodwin/pymagnum/files/8891574/magtest_2022-06-13_093925.txt)

The command was `magtest --device=/dev/ttyAMA0`

I think your problem may be the wiring. When I troubleshoot I assume anything and everything could be the problem.

you say you have plugged into a splitter with a BMK. does the BMK work as far as the Magnum system is concerned. If you swap BMK and the other cable, does BMK still work? That eliminates the splitter.
Are you certain the cable from the splitter to the DIN device is 4 wire. It must be 4 wire as the 2 outer wires are the magic wires. ;) Can you swap to another cable to test.

Are the wires from 2 &5 connected to A and B on the Waveshare device?

I'm running out of guesses.

BTW I tested with Version 2.0

**Reply by kkrankall** [reply]
Thank you Charles. I'm using the splitter that came with all of my magnum equipment and using the magnum supplied network cables.

Once I get home for the day I will reload the pi from scratch and try again. If that doesn't work I'll order up a new HAT or USB and test that. 

Looking forward to getting this up and running.

**Comment by CharlesGodwin** [comment, marked as answer]
All should be good. Just double check every connection is right and tight. It's got to be something. I find it's always the least suspicious item.

**Reply by CharlesGodwin** [reply]
Any luck?

**Reply by kkrankall** [reply]
Just received a new USB RS485 dongle and now I'm getting data.

```
magtest --device /dev/ttyUSB0
Magnum Test Version:2.0
Options:device='/dev/ttyUSB0', packets=50, timeout=0.005, trace=False, cleanpackets=True, log=False
Length:16 REMOTE_C  =>4000011F000C74B401003D1937286B01
Length:21 REMOTE_11 =>000528786405280600B4FF78001414000000000011
Length:18 BMK_81    =>81640B32FFF608990B8B00010B7F00020A01
Length:21 INVERTER  =>4000011F000C74B401003D1937286B010003025800
Length:21 REMOTE_A1 =>000528786405280600B4FF780014600090787878A1
Length:21 INVERTER  =>4000011F000C75B301003D1937286B010003025800
Length:21 REMOTE_D0 =>000528786405280600B4FF780014000000000000D0
Length:21 INVERTER  =>4000011F000C75B301003D1937286B010003025800
Length:21 REMOTE_C0 =>000528786405280600B4FF780014000000080000C0
/dev/ttyUSB0 module 'magnum' has no attribute 'magnumPT100'
Error detected attempting to read network data - test failed.
```

it says test failed but it looks like it read data. I was running a PT100 that I removed from the system and put back my Victron MTTPs. Not sure if that is causing a problem

**Comment by CharlesGodwin** [comment]
Your USB adapter works, but alas, you've uncovered a bug in my software (blush). I'm working on a fix.

**Comment by CharlesGodwin** [comment]
The bug is in `magtest`. I'm having a problem fixing it but try `magdump`, it should work. I'll have `magtest` working tomorrow, or maybe tonight. The output will be different, it will be JSON output.

**Reply by kkrankall** [reply]
No rush. I rolled back to version 1.1.3 and things are working

I'm a C# and Swift programer so this is somewhat new to me. I copied over your mqttlogger.py script and when I try to excute the script I get this error. The only thing I changed was the IP address of the broker to my HASS pi.

```
python3 mqttlogger.py
  File "/home/kkrankall/mqttlogger.py", line 77
    parser.add_argument("-d", "--device", nargs='*', action='append', default=[],
    ^
SyntaxError: invalid syntax
```

Do I need to import anything else to make this script work?

**Reply by CharlesGodwin** [reply]
I'll check it out. The examples are not supported. ;) But I'll fix it. your code is correct.

If you have the patience, I have a tentative fix for magtest

- download this file in a browser [URL](https://godwin.ca/wp-content/uploads/2022/06/pymagnum-2.0.1.zip)
if you're in a pi use `wget https://godwin.ca/wp-content/uploads/2022/06/pymagnum-2.0.1.zip`
- `sudo pip3 install pymagnum-2.0.1.zip`

run magtest again. if it works, great, i'll publish it. If not `sudo pip3 uninstall pymagnum`
**EDIT** I have deleted the zip file.

**Reply by kkrankall** [reply]
Thank you Charles. Yes you fixed it!

```
magtest --device=/dev/ttyUSB0
Magnum Test Version:2.0.1
Options:device='/dev/ttyUSB0', packets=50, timeout=0.005, trace=False, cleanpackets=True, log=False
Length:16 REMOTE_C  =>4000010F000A77B201003D1937276B01
Length:21 REMOTE_D0 =>000528786405280600B4FF780014000000000000D0
Length:21 INVERTER  =>4000010F000A77B201003D1937276B010002025800
Length:21 REMOTE_C0 =>000528786405280600B4FF780014000000100000C0
Length:21 INVERTER  =>40000110000A77B201003D1937276B010002025800
Length:23 UNKNOWN   =>000528786405280600B4FF780014041700000000009128
Length:21 INVERTER  =>40000110000A74B301003D1937276B010002025800
Length:21 REMOTE_80 =>000528786405280600B4FF78001404170000280080
Length:18 BMK_81    =>81640A96FFCE08990B8B00010B9000020A01
Length:16 REMOTE_C  =>40000110000A74B301003D1937276B01
Length:21 REMOTE_A0 =>000528786405280600B4FF780014041714007300A0
Length:21 INVERTER  =>40000110000A74B301003D1937276B010002025800
Length:21 REMOTE_11 =>000528786405280600B4FF78001414000000000011
Length:21 INVERTER  =>40000110000A74B301003D1937276B010002025800
Length:21 REMOTE_A1 =>000528786405280600B4FF780014600090787878A1
Length:21 INVERTER  =>40000110000874B301003D1937276B010002025800
Length:21 REMOTE_D0 =>000528786405280600B4FF780014000000000000D0
Length:16 REMOTE_C  =>40000110000A74B301003D1937276B01
Length:21 REMOTE_C0 =>000528786405280600B4FF780014000000080000C0
Length:18 BMK_81    =>81640A96FFD008990B8B00010B9000020A01
Length:21 INVERTER  =>40000110000A74B301003D1937276B010002025800
Length:21 REMOTE_00 =>000528786405280600B4FF78001404170000000000
Length:21 REMOTE_00 =>000110000A74B301003D1937276B01000202580000
Length:21 REMOTE_80 =>000528786405280600B4FF78001404170000280080
Length:21 INVERTER  =>40000110000A74B301003D1937276B010002025800
Length:21 REMOTE_A0 =>000528786405280600B4FF780014041714007300A0
Length:16 REMOTE_C  =>40000110000A74B401003D1937276B01
Length:21 REMOTE_11 =>000528786405280600B4FF78001414000000000011
Length:21 INVERTER  =>40000110000A74B401003D1937276B010002025800
Length:21 REMOTE_A1 =>000528786405280600B4FF780014600090787878A1
Length:18 BMK_81    =>81640A95FFD008990B8B00010B9000020A01
Packets:31 of 50 with 1 UNKNOWN, in 1.91 seconds
INVERTER Detected
REMOTE Detected
RTR not connected
BMK Detected
AGS not connected
PT100 not connected
ACLD not connected

```

**Reply by CharlesGodwin** [reply]
found the mqttlogger bug.

on line 76 add a `)` at end of line
on line 77 change `parser` to `logger`

python is column sensitive, so logger must start in column 1, the same column as parser. (I don't make the rules)

I'll publish 2.0.1 when I know it works.

PS, although I've learned more than 20 programming languages, C# and Swift are not on that list. ;)

**Reply by CharlesGodwin** [reply]
I'll publish tomorrow
Thanks for your help and patience

**Reply by kkrankall** [reply]
No thank you. Tweeked the mqttlogger to add credentials. 

Think things are working. Thank you for this amazing tool

**Comment by kkrankall** [comment]
Charles,

Thank you again. I updated the mqttlogger routines to stay connected to the broker to help speedup the refreshing of the data. Works like a charm. All of my data is now in Home Assistiant

<img width="459" alt="image" src="https://user-images.githubusercontent.com/15572661/174240627-52e7edab-f0c6-4f65-947c-dcb7126f840d.png">

**Reply by CharlesGodwin** [reply]
Enjoy
I'm still puzzled why the HAT didn't work. During your testing I installed a HAT and it worked for me. And I know of others using the HAT too.

Concerning the MQTT example. I wrote my code attempting to be bullet proof so it could recover from a temporary error. Optimization is always possible.

If you wish, share your code about authentication and I may be able to add it to the base example.

**Reply by kkrankall** [reply]
I think the HAT was bad. I tried it on 2 other pi's and it did the same thing. I'm going to clean up my code in the next few days and will share it for sure.

**Comment by kkrankall** [comment]
Here is the MQTT code I'm using. This is the first time I have anything in python so it's not pretty but majority of the code is yours still. I set my interval very sort and I'm filtering what devices I'm publishing. I also had a strange issues where once in a while it would show another inverter like a MMS model that would have a bunch of faults and inverter mode would be ??

The only major difference is this does not disconnected from the MQTT Broker and keeps on looping.

```
import json
import signal
import sys
import time
import uuid
from collections import OrderedDict
from datetime import datetime

import paho.mqtt.client as mqtt
from magnum.magnum import Magnum
from magnum.magparser import MagnumArgumentParser
from tzlocal import get_localzone


def sigint_handler(signal, frame):
    print('Interrupted. Shutting down.')
    sys.exit(0)

def on_connect(client, userdata, flags, rc):
 
    if rc == 0:
 
        print("Connected to broker")
 
        global Connected                #Use global variable
        Connected = True                #Signal connection 
 
    else:
 
        print("Connection failed")


signal.signal(signal.SIGINT, sigint_handler)
parser = MagnumArgumentParser(description="Magnum Data MQTT Publisher", fromfile_prefix_chars='@',
                              epilog="Refer to https://github.com/CharlesGodwin/pymagnum for details")
logger = parser.add_argument_group("MQTT publish")
reader = parser.add_argument_group("Magnum reader")
seldom = parser.add_argument_group("Seldom used")
logger.add_argument("-t", "--topic", default='magnum/',
                    help="Topic prefix (default: %(default)s)")
logger.add_argument("-b", "--broker", default='127.0.0.1:1883',
                    help="MQTT Broker address and (optional port)(default: %(default)s)")
logger.add_argument("-i", "--interval", default=10, type=int, dest='interval',
                    help="Interval, in seconds, between publishing (default: %(default)s)")
logger.add_argument("-d", "--device", nargs='*', action='append', default=[],
                    help="Serial device name (default: /dev/ttyUSB0). You can specify more than one.")
seldom.add_argument("--packets", default=50, type=int,
                    help="Number of packets to generate in reader (default: %(default)s)")
seldom.add_argument("--timeout", default=0.005, type=float,
                    help="Timeout for serial read (default: %(default)s)")
seldom.add_argument("--trace", action="store_true",
                    help="Add raw packet info to data (default: %(default)s)")
seldom.add_argument("--nocleanup", action="store_true", default=False, dest='cleanpackets',
                    help="Suppress clean up of unknown packets (default: False)")
args = parser.magnum_parse_args()
if args.interval < 10 or args.interval > (60*60):
    parser.error(
        "argument -i/--interval: must be between 10 seconds and 3600 (1 hour)")
if args.topic[-1] != "/":
    args.topic += "/"
print("Options:{}".format(str(args).replace("Namespace(", "").replace(")", "")))
magnumReaders = dict()
for device in args.device:
    try:
        magnumReader = Magnum(device=device, packets=args.packets, trace=args.trace,
                              timeout=args.timeout, cleanpackets=args.cleanpackets)
        magnumReader.getDevices()  # test read to see if all's good
        magnumReaders[magnumReader.getComm_Device()] = magnumReader
    except Exception as e:
        print("{0} {1}".format(device, str(e)))
if len(magnumReaders) == 0:
    print("Error: There are no usable devices connected.")
    exit(2)

print("Publishing to broker:{0} Every:{2} seconds. Using: {1} ".format(
    args.broker, list(magnumReaders.keys()), args.interval))
inverterModel = "MS4024PAE"  #I'm adding this to filter and inverter model that isn't mine
Connected = False
uuidstr = str(uuid.uuid1())
client = mqtt.Client(client_id=uuidstr, clean_session=False)
brokerinfo = args.broker.split(':')
if len(brokerinfo) == 1:
    brokerinfo.append(1883)
client.username_pw_set("ooooooo", password="xxxxxxxx")    #set username and password
client.on_connect= on_connect                      #attach function to callback
client.connect(brokerinfo[0], port=int(brokerinfo[1]))        #connect to broker 
client.loop_start()        #start the loop
while Connected != True:    #Wait for connection
    time.sleep(0.1)
try:
    while True:
        for comm_device, magnumReader in magnumReaders.items():
            try:
                devices = magnumReader.getDevices()
                print("Getting Devices")
                if len(devices) != 0:
                    data = OrderedDict()
                    now = int(time.time())
                    data["datetime"] = datetime.now(
                        get_localzone()).replace(microsecond=0).isoformat()
                    data['comm_device'] = comm_device
                    for device in devices:
                        topic = args.topic + device["device"].lower()
                        data["device"] = device["device"]
                        data["data"] = device["data"]
                        payload = json.dumps(
                            data, indent=None, ensure_ascii=True, allow_nan=True, separators=(',', ':'))
                        if device["device"].lower() == "inverter" and inverterModel in payload :
                            client.publish(topic, payload=payload)
                            print("{0}".format(topic))
                            print("{0}".format(payload))
                        if device["device"].lower() == "bmk":
                            client.publish(topic, payload=payload)
                            print("{0}".format(topic))
                            print("{0}".format(payload))
            except Exception as e:
                print("{0} {1}".format(comm_device, str(e)))
        time.sleep(args.interval)
 
except KeyboardInterrupt:
 
    client.disconnect()
    client.loop_stop()
```

**Comment by CharlesGodwin** [comment]
Thanks. I'll review and ask questions as needed. Not today though, I have chores to do first.

**Comment by CharlesGodwin** [comment]
I've reviewed your code.
I'm going to open some issues, instead of using discussion thread.
they will be:

- add support for user credential - should be straightforward
- add support for the 'loop'

As this program is an example, users are free to modify as they wish and updates to the package will not overwrite user's code.

If this was an issue, I would consider it closed.

thanks for the input
